### PR TITLE
support sharding  llama2-70b on more than 8 GPUs

### PR DESCRIPTION
**Pull request by wejoncy**
For now Llama2 has only 8 KV heads so we can't run Llama2-70b on more than 8 GPUs.

This PR is to support it.

### CODE REVIEW

File: vllm/model_executor/models/llama.py [hunk 1]
```diff
@@ -102,8 +102,9 @@ def __init__(
         assert self.total_num_heads % tp_size == 0
         self.num_heads = self.total_num_heads // tp_size
         self.total_num_kv_heads = num_kv_heads
-        assert self.total_num_kv_heads % tp_size == 0
-        self.num_kv_heads = self.total_num_kv_heads // tp_size
+        assert max(self.total_num_kv_heads, tp_size)% min(self.total_num_kv_heads, tp_size) == 0
```
Comment (zhuohan123) [inline vllm/model_executor/models/llama.py:105]: Could you comment on the logic here?

File: vllm/model_executor/models/llama.py [hunk 2]
```diff
@@ -102,8 +102,9 @@ def __init__(
...
+        self.num_kv_heads = max(1, self.total_num_kv_heads // tp_size)
+        repeat_num = max(1, tp_size//self.total_num_kv_heads)
```
Comment (zhuohan123) [inline vllm/model_executor/models/llama.py:107]: Nit: naming `repeat_num` -> `num_kv_heads_replicas`

File: vllm/model_executor/models/llama.py (hunk 1 above)
Comment (zhuohan123) [reply vllm/model_executor/models/llama.py:105]: A little bit longer but easier to understand:
```suggestion
        if self.total_num_kv_heads >= tp_size:
            assert self.total_num_kv_heads % tp_size == 0
        else:
            assert tp_size % self.total_num_kv_heads == 0
```

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
Thanks for your contribution! Please also run the formatter `format.sh` to fix the formatter issue.

**Comment by wejoncy** [comment]
Hi @zhuohan123 Thanks for your review.
Issues have been fixed.

**Comment by zhuohan123** [comment]
@wejoncy Can you enable the edit permission of this PR? I would like to submit some modifications and merge this PR.

**Comment by wejoncy** [comment]
Hi @zhuohan123 , It's enabled.

Since you created a nother PR, maybe I can close this one.

**Comment by esmeetu** [comment]
@wejoncy I tried understand this method. The main idea is that if i can't have enough memory with 8 gpus to run whole model with 8 kv keads, we put 8 kv heads onto 16 gpus using two replicas kv heads but split other attention layers as usual？ Only kv layers are duplicated, others are split without duplication. Am i right?

**Comment by wejoncy** [comment]
> @wejoncy I tried understand this method. The main idea is that if i have not enough memory with 8 gpus to run whole model with 8 kv keads, we can put 8 kv heads onto 16 gpus using two replicas kv heads but split other attention layers as usual？ Only kv layers are duplicated, others are split without duplication. Am i right?

Yes, exactly.

**Pull request closed without merging**